oeps/architectural-decisions/oep-0065-frontend-composability.rst

```diff
@@ -0,0 +1,486 @@
+------------------------------------
+
+With the creation of the `frontend-plugin-framework <https://github.com/openedx/frontend-plugin-framework>`_, we're starting to have the capability to share components across MFEs as plugins, either at build time (direct plugins) or runtime (iframe plugins)
+
+Direct plugins create some flexibility, but couple our repositories' builds together. Iframes are good for sandboxing and isolating code, but they're a very inefficient way to compose a UI, especially given the consistency and UX/DX concerns raised above. In a way, they exacerbate the problem even more. There's no way to do direct plugins via the MFE config API, since they rely on importing modules directly into the build.
+
+Specification
+*************
+
+Fundamentally, enabling runtime module loading and shared dependencies address the majority of these problems, so our approach centers on enabling these two things.
+
+Runtime Module Loading
+======================
+
+The capability to lazily load content from independently built modules into the page - without iframes - solves a few of the issues mentioned above. In particular, it gives us a way of composing UI elements from different MFEs/Domains dynamically at runtime without a "host" application needing to know anything about the "guest" at build-time. The two remain completely decoupled, save some shared runtime configuration.
+
+This capability has a high degree of overlap with the goals of the `frontend-plugin-framework <https://github.com/openedx/frontend-plugin-framework>`_ (FPF), and was the inspiration and approach for the original POC for that framework. We expect that the FPF would add a new plugin type - probably "module plugin" - which handles loading plugins via runtime module loading.
+
+Runtime module loading advantages
+---------------------------------
+
+- Reduces the frequency of full page refreshes. MFEs today are completely independent, so navigating between them means loading a completely new page (even if they share dependencies).
+- Improves composability across MFEs/domains. We have no way to show more than one MFE on the same page today except by using iframes or by creating hard dependencies between MFEs at build-time by extracting 'shared' code into a new library, like `frontend-component-header <https://github.com/openedx/frontend-component-header>`_, `frontend-component-footer <https://github.com/openedx/frontend-component-footer>`_, or `frontend-lib-content-components <https://github.com/openedx/frontend-lib-content-components>`_, which increases our dependency maintenance burden significantly.
+- Improves runtime extensiblity by allowing us to configure where an MFE's code should be loaded from, rather than needing to build it in to an app. It dovetails nicely with the FPF by providing us with a seamless, performant, and flexible way of extending our frontends without needing to rebuild the host.
+
+Shared Dependencies
+===================
+
```

Comment:
We can utilize `Shared Dependencies` without necessary using module federation right?.

Reply:
There are other ways to do it, yes.  But MF also provides logic to deal with incompatibilities between versions that are already loaded and those needed by dynamically loaded modules, which is pretty nice.

Reply:
@sarina, are you saying you'd like to see something here about module federation's logic for dealing with incompatibilities, or about how we can have shared deps without using MF?  

In either case, at this point in the OEP was trying not to get into the implementation (module federation) too much, reserving that for the 'refererence implementation' section.

Reply:
No action needed here. I was bookmarking the comment but I think I got enough of the gist of how incompatibilites work later on.